cudamapper/src/chainer_utils.cuh

```diff
@@ -0,0 +1,149 @@
+                            device_buffer<int32_t>& unrolled_anchor_chains,
+                            device_buffer<int32_t>& anchor_chain_starts,
+                            int32_t num_overlaps,
+                            int32_t& num_total_anchors,
+                            DefaultDeviceAllocator& _allocator,
+                            cudaStream_t& _cuda_stream);
+
+///
+/// \brief Calculate the anchors chains used to produce each overlap in the
+/// overlap array for anchors chained by backtrace_anchors_to_overlaps.
+///
+/// \param overlaps An array of overlaps.
+/// \param anchors The array of anchors used to generate overlaps.
+/// \param select_mask An array of bools, used to mask overlaps from output.
+/// \param predecessors The predecessors array from anchor chaining.
+/// \param anchor_chains An array (allocated by allocate_anchor_chains) which will hold
+/// the indices of anchors within each chain.
+/// \param anchor_chain_starts An array which holds the indices of the first anchor for
+/// each overlap in the overlaps array.
+/// \param num_overlaps The number of overlaps in the overlaps array
+/// \param check_mask A boolean. If true, only overlaps where select_mask is true
+/// will have their anchor chains calculated.
+/// \return __global__
+__global__ void output_overlap_chains_by_backtrace(const Overlap* overlaps,
+                                                   const Anchor* anchors,
+                                                   const bool* select_mask,
+                                                   const int32_t* predecessors,
+                                                   int32_t* anchor_chains,
+                                                   int32_t* anchor_chain_starts,
+                                                   int32_t num_overlaps,
+                                                   bool check_mask);
+///
+/// \brief Calculate the anchors chains used to produce each overlap in the
+/// overlap array for anchors chained by RLE.
+///
+/// \param overlaps An array of overlaps.
+/// \param anchors The array of anchors used to generate overlaps.
+/// \param chain_starts An array which holds the indices of the first anchor for
+/// each overlap in the overlaps array.
+/// \param chain_lengths An array which holds the length of each run of anchors, corresponding to
+/// the chain_starts array.
+/// \param anchor_chains An array (allocated by allocate_anchor_chains) which will hold
+/// the indices of anchors within each chain.
+/// \param anchor_chain_starts An array which holds the indices of the first anchor for
+/// each overlap in the overlaps array.
+/// \param num_overlaps The number of overlaps in the overlaps array.
+/// \return __global__
+__global__ void output_overlap_chains_by_RLE(const Overlap* overlaps,
+                                             const Anchor* anchors,
+                                             const int32_t* chain_starts,
```

Comment:
How confident are we that we won't have more than 2^31 anchors? An example I usually use regularly has 300M anchors

Reply:
Not terribly confident actually. I've refactored it to an unsigned int.

Reply:
I was going for 32/64-bit discussion and not 31/32

Reply:
That does make sense - raised to 64 bits in 781c2fd2.